apps/api/src/services/data-provider.service.ts

```diff
@@ -132,73 +130,45 @@ export class DataProviderService implements DataProviderInterface {
       [symbol: string]: { [date: string]: IDataProviderHistoricalResponse };
     } = {};
 
+    const promises: Promise<{
+      data: { [date: string]: IDataProviderHistoricalResponse };
+      symbol: string;
+    }>[] = [];
     for (const { dataSource, symbol } of aDataGatheringItems) {
-      switch (dataSource) {
-        case DataSource.ALPHA_VANTAGE: {
-          if (this.configurationService.get('ALPHA_VANTAGE_API_KEY')) {
-            const data = await this.alphaVantageService.getHistorical(
-              [symbol],
-              undefined,
-              from,
-              to
-            );
-            result[symbol] = data?.[symbol];
-          }
-          break;
-        }
-        case DataSource.GHOSTFOLIO: {
-          if (isGhostfolioScraperApiSymbol(symbol)) {
-            const data = await this.ghostfolioScraperApiService.getHistorical(
-              [symbol],
-              undefined,
-              from,
-              to
-            );
-            result[symbol] = data?.[symbol];
-          }
-          break;
-        }
-        case DataSource.RAKUTEN: {
-          if (
-            isRakutenRapidApiSymbol(symbol) &&
-            this.configurationService.get('RAKUTEN_RAPID_API_KEY')
-          ) {
-            const data = await this.rakutenRapidApiService.getHistorical(
-              [symbol],
-              undefined,
-              from,
-              to
-            );
-            result[symbol] = data?.[symbol];
-          }
-          break;
-        }
-        case DataSource.YAHOO: {
-          const data = await this.yahooFinanceService.getHistorical(
-            [symbol],
-            undefined,
-            from,
-            to
-          );
-          result[symbol] = data?.[symbol];
-          break;
-        }
+      const dataProvider = this.getDataProvider(dataSource);
+      if (dataProvider.hasHistoricalData(symbol)) {
+        promises.push(
+          dataProvider
+            .getHistorical([symbol], undefined, from, to)
+            .then((data) => ({ data: data?.[symbol], symbol }))
```

Comment:
Can you refactor to `async` / `await`?

Reply:
No, `async` / `await` doesn't support multiple promises at the same time - if you want the unoptimized version, you just need to cherry-pick the first commit